File: 322-coin-change/coin-change.java
class Solution {
    int[] dp;
    int coinChange(int[] coins, int amount) {
        dp = new int[amount + 1];
        Arrays.fill(dp , -1);
        int ans = coinCount(coins, amount);
        return ans == Integer.MAX_VALUE ? -1 : ans;
    }

    int coinCount(int[] coins, int amount) {
       if(amount == 0) return 0;

       if(amount <0) return Integer.MAX_VALUE;

        if(dp[amount] != -1) return dp[amount];
       int min_coin = Integer.MAX_VALUE;

       for(int i = 0; i<coins.length;i++) {
            int ans = coinCount(coins, amount - coins[i]);
            if(ans != Integer.MAX_VALUE)
                min_coin = Math.min(min_coin, 1 + ans);
       }
       return dp[amount] = min_coin;
    }
}

File: 322-coin-change/coin-change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        
        #---------------------------- Bottom Up ---------------------------#

        mem = [amount + 1] * (amount + 1)
        mem[0] = 0

        for a in range(1, amount + 1):
            for coin in coins:
                if a - coin >=0:
                    mem[a] = min(mem[a], 1 + mem[a - coin])

        return mem[amount] if mem[amount] != amount + 1 else -1


        #---------------------------- Top Down ---------------------------#
        # coins.sort()
        # memo = {}
        # def rec(target):
        #     if target == 0:
        #         return 0

        #     if target in memo:
        #         return memo[target]

        #     minRes = float("inf")

        #     for i in coins:
        #         if target - i >=0:
        #             minRes = min(minRes, 1 + rec(target - i))

        #     memo[target] = minRes
        #     return memo[target]

        # res = rec(amount)
        # return res if res != float("inf") else -1
            
            
